Mathlib/Order/RelSeries.lean

```diff
@@ -0,0 +1,261 @@
+
+lemma rel_or_eq_of_le [IsTrans α r] (x : RelSeries r) {i j : Fin (x.length + 1)} (h : i ≤ j) :
+    r (x i) (x j) ∨ x i = x j :=
+  (le_iff_lt_or_eq.mp h).by_cases (Or.intro_left _ $ x.rel_of_lt .) (Or.intro_right _ $ . ▸ rfl)
+
+/--
+Given two relations `r, s` on `α` such that `r ≤ s`, any relation series of `r` induces a relation
+series of `s`
+-/
+@[simps!]
+def OfLE (x : RelSeries r) {s : Rel α α} (h : r ≤ s) : RelSeries s where
+  length := x.length
+  toFun := x
+  step := fun _ => h _ _ <| x.step _
+
+lemma ofLE_length (x : RelSeries r) {s : Rel α α} (h : r ≤ s) :
+    (x.OfLE h).length = x.length := rfl
+
+lemma coe_ofLE (x : RelSeries r) {s : Rel α α} (h : r ≤ s) :
+  (x.OfLE h : _ → _) = x := rfl
+
+/-- Every relation series gives a list -/
+abbrev toList (x : RelSeries r) : List α := List.ofFn x
+
+lemma toList_chain' (x : RelSeries r) : x.toList.Chain' r := by
+  rw [List.chain'_iff_get]
+  intros i h
+  have h' : i < x.length := by simpa [List.length_ofFn] using h
+  convert x.step ⟨i, h'⟩ <;>
+  · rw [List.get_ofFn]
+    congr 1
+
+lemma toList_ne_empty (x : RelSeries r) : x.toList ≠ ∅ := fun m =>
+  List.eq_nil_iff_forall_not_mem.mp m (x 0) <| (List.mem_ofFn _ _).mpr ⟨_, rfl⟩
+
+/-- Every nonempty list satisfying the chain condition gives a relation series-/
+@[simps]
+def fromListChain' (x : List α) (x_ne_empty : x ≠ ∅) (hx : x.Chain' r) : RelSeries r where
+  length := x.length.pred
+  toFun := x.get ∘ Fin.cast (Nat.succ_pred_eq_of_pos <| List.length_pos.mpr x_ne_empty)
+  step := fun i => List.chain'_iff_get.mp hx i i.2
+
+/-- Relation series of `r` and nonempty list of `α` satisfying `r`-chain condition bijectively
+corresponds to each other.-/
+protected def Equiv : RelSeries r ≃ {x : List α | x ≠ ∅ ∧ x.Chain' r} where
+  toFun := fun x => ⟨_, x.toList_ne_empty, x.toList_chain'⟩
+  invFun := fun x => fromListChain' _ x.2.1 x.2.2
+  left_inv := fun x => ext (by dsimp; rw [List.length_ofFn, Nat.pred_succ]) <| by
+    ext f
+    simp only [List.empty_eq, ne_eq, Set.coe_setOf, Set.mem_setOf_eq, fromListChain'_toFun,
+      Function.comp_apply, List.get_ofFn, fromListChain'_length]
+    rfl
+  right_inv := by
+    intro x
+    refine Subtype.ext (List.ext_get ?_ <| fun n hn1 hn2 => ?_)
+    · dsimp
+      rw [List.length_ofFn, fromListChain'_length, ←Nat.succ_eq_add_one, Nat.succ_pred_eq_of_pos]
+      rw [List.length_pos]
+      exact x.2.1
+    · rw [List.get_ofFn, fromListChain'_toFun, Function.comp_apply]
+      congr
+
+-- TODO : build a similar bijection between `RelSeries α` and `Quiver.Path`
+
+end RelSeries
+
+namespace Rel
+
+/-- a relation `r` is said to be finite dimensional iff there is a relation series of `r` with the
+  maximum length. -/
+class FiniteDimensional where
+  /-- the longest relation series-/
+  longestRelSeries : RelSeries r
+  /-- the longest relation series is longer than any other series -/
+  is_longest : ∀ (y : RelSeries r), y.length ≤ longestRelSeries.length
```

Comment:
Is there a reason that you want to keep track of a particular longest RelSeries? Notice that data held by a class is supposed to be canonical structure on a type rather than an arbitrarily chosen distinguished RelSeries. Personally I'd just define this class to consists of a single field: `exists_longest : ∃ longest : RelSeries r, ∀ y : RelSeries r, y.length ≤ longest.length`. This way the class would be Prop-valued and therefore couldn't be involved in any diamonds, and also conform to the design of [Finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finite/Defs.html#Finite), [Module.Finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Finiteness.html#Module.Finite) / [FiniteDimensional](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FiniteDimensional.html#FiniteDimensional), [Algebra.FiniteType](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteType.html#Algebra.FiniteType), etc.

Reply:
I was trying to avoid adding non-computable, but I guess I don't really care about computability issues. So I will change it to a `Prop` valued one

Reply:
changed